File: backend/src/main/kotlin/hu/bme/sch/cmsch/component/profile/ProfileView.kt
package hu.bme.sch.cmsch.component.profile

import com.fasterxml.jackson.annotation.JsonView
import hu.bme.sch.cmsch.component.debt.DebtDto
import hu.bme.sch.cmsch.component.leaderboard.LeaderBoardEntry
import hu.bme.sch.cmsch.component.token.TokenDto
import hu.bme.sch.cmsch.dto.*
import hu.bme.sch.cmsch.model.GuildType
import hu.bme.sch.cmsch.model.MajorType
import hu.bme.sch.cmsch.model.RoleType

data class ProfileView(
    @field:JsonView(FullDetails::class)
    val loggedIn: Boolean = false,

    @field:JsonView(FullDetails::class)
    val fullName: String? = null,

    @field:JsonView(FullDetails::class)
    val groupName: String? = null,

    @field:JsonView(FullDetails::class)
    val role: RoleType = RoleType.GUEST,

    @field:JsonView(FullDetails::class)
    val alias: String? = null,

    @field:JsonView(FullDetails::class)
    val guild: GuildType? = null,

    @field:JsonView(FullDetails::class)
    val email: String? = null,

    @field:JsonView(FullDetails::class)
    val neptun: String? = null,

    @field:JsonView(FullDetails::class)
    val cmschId: String? = null,

    @field:JsonView(FullDetails::class)
    val major: MajorType? = null,

    @field:JsonView(FullDetails::class)
    val groupLeaders: List<GroupLeaderDto>? = null,

    @field:JsonView(FullDetails::class)
    val groupSelectionAllowed: Boolean = false,

    @field:JsonView(FullDetails::class)
    val availableGroups: Map<Int, String>? = null,

    @field:JsonView(FullDetails::class)
    val fallbackGroup: Int? = null,

    @field:JsonView(FullDetails::class)
    val tokens: List<TokenDto>? = null,

    @field:JsonView(FullDetails::class)
    val collectedTokenCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val totalTokenCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val minTokenToComplete: Int? = null,

    @field:JsonView(FullDetails::class)
    val totalRiddleCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val completedRiddleCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val totalTaskCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val submittedTaskCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val completedTaskCount: Int? = null,

    @field:JsonView(FullDetails::class)
    val profileIsComplete: Boolean? = null,

    @field:JsonView(FullDetails::class)
    val incompleteTasks: List<String>? = null,

    @field:JsonView(FullDetails::class)
    val locations: List<GroupMemberLocationDto>? = null,

    @field:JsonView(FullDetails::class)
    val debts: List<DebtDto>? = null,

    @field:JsonView(FullDetails::class)
    val leaderboard: List<LeaderBoardEntry>? = null,

    @field:JsonView(FullDetails::class)
    val groupMessage: String? = null,

    @field:JsonView(FullDetails::class)
    val userMessage: String? = null,
)

File: frontend/src/util/views/profile.view.ts
import type { DebtView } from './debt.view'
import type { GroupLeaderView } from './groupLeader.view'
import type { GroupMemberLocationView } from './groupMemberLocation.view'
import type { TokenView } from './token.view'
import type { TopListAbstractEntryView } from './toplistAbstractEntry.view'

export interface ProfileView {
  cmschId?: string
  loggedIn: boolean
  fullName?: string
  alias?: string
  email?: string
  neptun?: string
  guild?: keyof typeof GuildType
  major?: MajorType

  groupName?: string
  groupLeaders?: GroupLeaderView[]
  groupSelectionAllowed: boolean
  availableGroups?: Record<number, string>
  fallbackGroup?: number

  role: keyof typeof RoleType

  tokens?: TokenView[]
  minTokenToComplete?: number
  totalTokenCount?: number
  collectedTokenCount?: number

  totalRiddleCount?: number
  completedRiddleCount?: number

  totalTaskCount?: number
  submittedTaskCount?: number
  completedTaskCount?: number

  raceStats?: RaceStatsView

  locations?: GroupMemberLocationView[]
  debts?: DebtView[]
  leaderboard?: TopListAbstractEntryView[]

  profileIsComplete?: boolean
  incompleteTasks?: string[]

  groupMessage?: string
  userMessage?: string
}

export interface RaceStatsView {
  bestTime: number
  placement: number
  timesParticipated: number
  averageTime: number
  deviation: number
  kCaloriesPerSecond: number
}

//cannot compare roles if the enums values are strings use the RoleType[role] syntax
export const RoleType = {
  GUEST: 0,
  BASIC: 1,
  ATTENDEE: 10,
  PRIVILEGED: 11,
  STAFF: 100,
  ADMIN: 200,
  SUPERUSER: 500
}
export type RoleType = (typeof RoleType)[keyof typeof RoleType]
export type RoleTypeString = keyof typeof RoleType

export const RoleTypeString: { [k: string]: RoleTypeString } = {
  GUEST: 'GUEST',
  BASIC: 'BASIC',
  ATTENDEE: 'ATTENDEE',
  PRIVILEGED: 'PRIVILEGED',
  STAFF: 'STAFF',
  ADMIN: 'ADMIN',
  SUPERUSER: 'SUPERUSER'
}

export const GuildType = {
  UNKNOWN: 'Nincs',
  RED: 'Piros',
  BLACK: 'Fekete',
  BLUE: 'Kék',
  WHITE: 'Fehér',
  YELLOW: 'Sárga',
  PURPLE: 'Lila'
}
export type GuildType = (typeof GuildType)[keyof typeof GuildType]

export const MajorType = {
  UNKNOWN: 'UNKNOWN',
  IT: 'IT',
  EE: 'EE',
  BPROF: 'BPROF'
}
export type MajorType = (typeof MajorType)[keyof typeof MajorType]
